### ISSUE TITLE
GPULightmapper leaks light in situations where the CPU lightmapper wouldn't

### ISSUE THREAD

**Issue by jcostello** [godotengine/godot#51638]
### Godot version

v4.0.dev.calinou [6f043f7a1]

### System information

Mx Linux (Debian Based)

### Issue description

The lightmapper leaks even with thick walls.

![image](https://user-images.githubusercontent.com/1168582/129402395-5c8265fa-b0eb-4d2c-8118-7f3df6b3b8a2.png)

### Steps to reproduce

Use the minimal reproduction project and bake the lightmap

### Minimal reproduction project

[GPULightmapLeak.zip](https://github.com/godotengine/godot/files/6984512/GPULightmapLeak.zip)

**Comment by JFonS** [comment]
I did some testing on this, direct light seems to be correct and without leaks, but the first bounce breaks a number of things. I'm still not sure why. It may be some weird geometry or maybe something is wrong in the way we handle hits against backfaces, but there's definitely something going on that needs more investigation.

**Comment by jcostello** [comment]
I test it out with a simpler geometry (a cube with thick walls and a window, all conected) and the bleeding still happens

**Comment by williamd67** [comment]
I found one issue that GLULightmapper's triangles and bounding-boxes arrays became out-of-sync. This causes issues when the incorrect bounding-boxes were used to test the triangle in `trace_ray`. I fixed that issue in #52570. This does not fix the leaking, yet it does improves some other side-effects.

### Before
![Screenshot from 2021-09-11 12-51-39](https://user-images.githubusercontent.com/10560119/132947376-47211a98-78ff-4304-bf67-5905afa6eac6.png)
![Screenshot from 2021-09-11 12-55-12](https://user-images.githubusercontent.com/10560119/132947386-f250fc3c-162d-4d0a-ab11-d3cb166fccf3.png)
### After
![Screenshot from 2021-09-11 13-01-22](https://user-images.githubusercontent.com/10560119/132947392-26d85931-68d2-4262-aeb1-54069eee49e7.png)
![Screenshot from 2021-09-11 12-58-53](https://user-images.githubusercontent.com/10560119/132947393-e0201df1-f2d5-42e9-8c61-7ebff790d966.png)
![Screenshot from 2021-09-11 13-01-49](https://user-images.githubusercontent.com/10560119/132947413-b06d9096-c629-4914-a60b-3fa782a21100.png)

**Comment by jcostello** [comment]
Nice, it improved the other side effects. Any idea why the leaking might happen? I suspect that is because the UVs because even if you bake lighting around  2 boxes, the seams got some leaks

**Comment by williamd67** [comment]
I found another issue: rays that hit back-face triangles where counted in the bounce lighting calculations. Normally this was not the case, except for edges of planes where rays can get 'behind' the scene. This is fixed in #52618.

This did not solve the leaking completely, yet I found another setting that combined with this back-face exclusion fixed it. Changing the `Lightmap Scale` of the `Global Illumination` of the `Hut-mesh` from `1x` to `2x` removed all the leakage. Also to get a better effect I recommend to change the lightmap `quality` to `high`.

### After
![Screenshot from 2021-09-13 00-01-39](https://user-images.githubusercontent.com/10560119/133004622-17fd970e-c03f-4e9c-9b68-1a2fb348aa2a.png)
![Screenshot from 2021-09-12 23-42-25](https://user-images.githubusercontent.com/10560119/133004623-7eefa0ed-697d-4fe3-88a2-2170e1ea4988.png)

**Comment by jcostello** [comment]
I tried a similar model in Unity and I found that intercepting geometry is not well handled so as you said, scaling the mesh in the lightmap helped in Unity too. I tried it on Godot and helped but still and issue where the seams meet. So upscaling the mesh in the lightmap helps with intercepting geometry but still an issue with the seams.
In your after you can still see artifacts in the edges of the cube and in the room's corner.

**Comment by jcostello** [comment]
One thing we can try is to have a cube with custom uv2 and with generated uv2s. The custom  uv2 should have the problem only where the T seams meet. The generated uv2 will have in all the edges due to have every face with seems

**Comment by williamd67** [comment]
> In your after you can still see artifacts in the edges of the cube and in the room's corner.

Just to be sure, I have two fixes and two afters. In the second after there are no issues (lightmap is scaled). Although when you use `1x` it still does show issues. I plan to investigate further.

**Comment by jcostello** [comment]
Yes. Just to clarify, let me point out where I see the artifacts

Edge of the box
![image](https://user-images.githubusercontent.com/1168582/133127887-8efabcad-730f-4b87-93c1-6c68e1c27ccb.png)

Corner of the room.
![image](https://user-images.githubusercontent.com/1168582/133127958-bc1afb4f-2da0-4cf7-b0d9-934f07e40286.png)

If you need I can create another scenario to point this issues better

**Comment by williamd67** [comment]
I will have a look and another scenario might help. As you mentioned this could be related to the seaming phase which I did not investigate yet.

**Comment by williamd67** [comment]
I think the original issue where light is leaking through edges is fixed by #52618. I will continue to investigate the other artifacts that have been mentioned.

**Comment by JFonS** [comment]
I haven't followed the issue too closely, but the work by @williamd67 should help get rid of all the obvious leaks. @jcostello did you have a chance to test #52618? 

Something that isn't clear to me yet is whether the UV2s are imported along with the model or generated by Godot at import. Keep in mind that the various UV2 islands need to be separated by at least a 2 pixel gap, otherwise there's nothing the lightmapper can do to prevent light leaks due to bilinear filtering. Godot's unrwapper should already handle this, but we don't have any way to enforce this for imported UV2s.

**Comment by jcostello** [comment]
At first wtih #52618 I see new artifacts in this PR. I will create 2 more specific scenarios to upload so you guys can try

@JFonS @williamd67 

![image](https://user-images.githubusercontent.com/1168582/133994829-5c5704b3-d0c9-48bd-94fc-9e84b1287615.png)

**Comment by jcostello** [comment]
@JFonS @williamd67 I created 4 cases where I see problems with the light mapper. Here is the project you can see each case in the scene folder.

https://drive.google.com/file/d/1XbWk4HVYI-fH7PyzetGVKmUQ8O5VoEOq/view?usp=sharing

Test1.tscn - Hollow cube (solid) with omni light on it.
![image](https://user-images.githubusercontent.com/1168582/134010550-18d83ce3-6f27-441e-a3a8-48f9c8277b32.png)
Seems that the leaks are not occurring but there is a shadow on the seams. Its an intended AO?

Test2.tscn - Cubes with spotlight
![image](https://user-images.githubusercontent.com/1168582/134010829-3966edea-a430-4ebc-8d8d-56122c3df162.png)
Dark spot in the middle cube (causes by the last bounce?). Also white light in the edges.

![image](https://user-images.githubusercontent.com/1168582/134010950-0b887248-12df-4bb7-b002-eb727d18c775.png)
When moving the camera, a white square appears in the left cube's corner. Maybe not related with the lightmapper.

Test3.tscn - Hut model with omnilight in it.
This model has a lot of intersecting geometry. Not the best geometry but it will happen in games and some issues should be fixed.

![image](https://user-images.githubusercontent.com/1168582/134011150-d248efaf-6f4d-4c15-be48-c97e2f2f2d87.png)
Darker square inside the hut.

![image](https://user-images.githubusercontent.com/1168582/134011197-7d819a22-eb0c-4abe-9947-38d4540b3ce6.png)
Artifacts outside when using `x8` in the mesh scale inside the lightmap and `medium` quality in the lightmap node

Test4.tscn - complex model
![image](https://user-images.githubusercontent.com/1168582/134011479-5b303880-7527-4b21-8f7d-bceafa069497.png)
Leaking thought the model (doesn't seems to be a bounce)

**Comment by williamd67** [comment]
Thanks @jcostello, this is exactly what I need to test more and further. I will take your model and try to find and fix the issues (or explain why they happen).

**Comment by jcostello** [comment]
@williamd67 no problem. Let me know if I can help you testing your progress in any way :muscle:

**Comment by jcostello** [comment]
I retested this on master and seems that most of the things were fixed. Maybe I downloaded your PR @williamd67 but I didn't change the branch.

Anyway with the first test case, when baking at medium quality (3 bounces, mesh x1) the strange AO disappear (which is how it would work)

![image](https://user-images.githubusercontent.com/1168582/134521487-7080c4ec-bf92-4da0-a0df-a30ed8c58b83.png)

But baking it at ultru (3 bounces, mesh x1) it has an strange blob.

![image](https://user-images.githubusercontent.com/1168582/134521693-7fb154a5-2c95-4c7a-97cf-0d5fa3fcbea3.png)

Happens the same with x1, x2, x4 and x8 in the mesh size.

**Comment by jcostello** [comment]
With the hut example. The inside looks good as it should the outside still looks with artifacts but its fixed if baked at ultra.

Medium
![image](https://user-images.githubusercontent.com/1168582/134522367-58b3f262-f943-47e1-9208-4a454e02922e.png)


Ultra
![image](https://user-images.githubusercontent.com/1168582/134522040-4619d02f-cf13-42ed-a499-dff77c7a8d7c.png)

**Comment by jcostello** [comment]
scene 4 still have leaks but provably its because the geometry of the mesh

![image](https://user-images.githubusercontent.com/1168582/134524449-745c0524-ba5a-446a-b738-7ac8d15ebb6b.png)

**Comment by williamd67** [comment]
I found that the background color leaked, which I fixed in #53081. This PR fixes the test1.scn which still showed some background color leaking when there is no dynamic light.

This PR will not solve the other issues you found. I will continue to investigate those.

**Comment by jcostello** [comment]
Awesome. Great job @williamd67 . You are the man

**Comment by williamd67** [comment]
The above PR fixes the blob in scene 1 (baking at ultra, 3 bounces, 1x mesh).
